### PR TITLE
explain: add joins

### PR DESCRIPTION
**Pull request by CharLotteiu**
### What is changed, added or deleted? (Required)

Add `explain-joins.md`. This PR does not add this page to the TOC.

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs/pull/3922
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Have version specific changes 
- [ ] Might cause conflicts

### CODE REVIEW

File: explain-joins.md [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+6 rows in set (0.00 sec)
+```
+
+Index Join 算法对内存消耗较小，但如果需要执行大量探查操作，运行速度可能会慢于其他 Join 算法。以下面这条查询语句为例：
+
+```sql
+SELECT * FROM t1 INNER JOIN t2 ON t1.id=t2.t1_id WHERE t1.pad1 = 'value' and t2.pad1='value';
+```
+
+在 Inner Join 操作中，TiDB 会先执行 Join Reorder 算法，所以不能确定会先读取 `t1` 还是 `t2` 。假设 TiDB 先读取了 `t1` 来构建 Build 端，那么 TiDB 会在探查 `t2` 前先根据谓词 `t1.col = 'value'` 筛选数据，但接下来每次探查 `t2` 时都要应用谓词 `t2.col='value'`。所以对于这条语句，Index Join 算法可能不如其他 Join 算法高效。
```
Comment (TomShawn) [inline explain-joins.md:72]:
```suggestion
在 Inner Join 操作中，TiDB 会先执行 Join Reorder 算法，所以不能确定会先读取 `t1` 还是 `t2`。假设 TiDB 先读取了 `t1` 来构建 Build 端，那么 TiDB 会在探查 `t2` 前先根据谓词 `t1.col = 'value'` 筛选数据，但接下来每次探查 `t2` 时都要应用谓词 `t2.col='value'`。所以对于这条语句，Index Join 算法可能不如其他 Join 算法高效。
```

File: explain-joins.md [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+6 rows in set (0.36 sec)
+```
+
+> **注意：**
+>
+> 在上方示例中，SQL 优化器之所以选择了性能较差的 Hash Join 算法，而不是 Index Join 算法，原因在于查询优化是一个 [NP 完全问题](https://zh.wikipedia.org/wiki/NP%E5%AE%8C%E5%85%A8)，可能会选择不太理想的计划。如果需要频繁调用这个查询，建议通过[执行计划管理](/sql-plan-management.md)的方式将 Hint 与 SQL 语句绑定，这样要比在发送给 TiDB 的 SQL 语句中插入 Hint 更容易管理。
+
+### Index Join 相关算法
+
+如果使用 Hint [`INL_JOIN`](/optimizer-hints.md#inl_joint1_name--tl_name-) 进行 Index Join 操作，TiDB 会在连接外表之前创建一个中间结果的 Hash Table。TiDB 同样也支持使用 Hint [`INL_HASH_JOIN`](/optimizer-hints.md#inl_hash_join) 在外表上建 Hash Table。而如果内表列集合与外表的相匹配，则可以应用 Hint [`INL_MERGE_JOIN`](/optimizer-hints.md#inl_merge_join)。以上所列 Index Join 相关算法都由 SQL 优化器自动选择。
```
Comment (TomShawn) [inline explain-joins.md:185]:
```suggestion
如果使用 Hint [`INL_JOIN`](/optimizer-hints.md#inl_joint1_name--tl_name-) 进行 Index Join 操作，TiDB 会在连接外表之前创建一个中间结果的 Hash Table。TiDB 同样也支持使用 Hint [`INL_HASH_JOIN`](/optimizer-hints.md#inl_hash_join) 在外表上建 Hash Table。而如果内表列集合与外表的相匹配，则可以应用 Hint [`INL_MERGE_JOIN`](/optimizer-hints.md#inl_merge_join)。以上所述的 Index Join 相关算法都由 SQL 优化器自动选择。
```

File: explain-joins.md [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+
+### 配置
+
+Index Join 算法的性能受以下系统变量影响：
+
+* [`tidb_index_join_batch_size`](/system-variables.md#tidb_index_join_batch_size) (默认值：`25000`) - `index lookup join` 操作的 batch 大小。
+* [`tidb_index_lookup_join_concurrency`](/system-variables.md#tidb_index_lookup_join_concurrency) (默认值：`4`) - 可以并发执行的 index lookup 任务数。
```
Comment (TomShawn) [inline explain-joins.md:192]:
```suggestion
* [`tidb_index_join_batch_size`](/system-variables.md#tidb_index_join_batch_size)（默认值：`25000`）- `index lookup join` 操作的 batch 大小。
* [`tidb_index_lookup_join_concurrency`](/system-variables.md#tidb_index_lookup_join_concurrency)（默认值：`4`）- 可以并发执行的 index lookup 任务数。
```

File: explain-joins.md [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+| HashJoin_27                 | 142020.00 | root      |               | inner join, equal:[eq(test.t1.id, test.t2.id)] |
+| ├─TableReader_29(Build)     | 142020.00 | root      |               | data:TableFullScan_28                          |
+| │ └─TableFullScan_28        | 142020.00 | cop[tikv] | table:t1      | keep order:false                               |
+| └─TableReader_31(Probe)     | 180000.00 | root      |               | data:TableFullScan_30                          |
+|   └─TableFullScan_30        | 180000.00 | cop[tikv] | table:t2      | keep order:false                               |
++-----------------------------+-----------+-----------+---------------+------------------------------------------------+
+5 rows in set (0.00 sec)
+```
+
+TiDB 会按照以下操作步骤执行 `HashJoin_27` 算子：
```
Comment (TomShawn) [inline explain-joins.md:219]:
```suggestion
TiDB 会按照以下顺序执行 `HashJoin_27` 算子：
```

File: explain-joins.md [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+2. 根据缓存数据在 Build 端构造一个 Hash Table。
+3. 读取 Probe 端的数据。
+4. 使用 Probe 端的数据来探查 Hash Table。
+5. 将符合条件的结果返回给用户。
+
+`EXPLAIN` 返回结果中的 `operator info` 一列记录了 `HashJoin_27` 的其他信息，包括该查询是 Inner Join 还是 Outer Join 以及 Join 的条件是什么等。在上面给出的示例中，该查询为 Inner Join，Join 条件是 `equal:[eq(test.t1.id, test.t2.id)]`，与查询语句中的 `WHERE t1.id = t2.id` 部分对应。下面例子中其他几个 Join 算子的 operator info 和此处类似。
+
+### 运行数据
+
+如果在执行操作时，内存使用超过了 [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query) 规定的值（默认为 1GB），且 `oom-use-tmp-storage` 的值为 `true` （默认为真），那么 TiDB 会尝试使用临时存储，在磁盘上创建 Hash Join 的 Build 端。`EXPLAIN ANALYZE` 返回结果中的 `execution info` 一栏记录了有关内存使用情况等运行数据。下面的例子展示了 `tidb_mem_quota_query` 的值分别设为 1GB（默认）及 500MB 时，`EXPLAIN ANALYZE` 的返回结果（当内存配额设为 500MB 时，磁盘用作临时存储区）：
```
Comment (TomShawn) [inline explain-joins.md:231]:
```suggestion
如果在执行操作时，内存使用超过了 [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query) 规定的值（默认为 1GB），且 `oom-use-tmp-storage` 的值为 `true` （默认为 `true`），那么 TiDB 会尝试使用临时存储，在磁盘上创建 Hash Join 的 Build 端。`EXPLAIN ANALYZE` 返回结果中的 `execution info` 一栏记录了有关内存使用情况等运行数据。下面的例子展示了 `tidb_mem_quota_query` 的值分别设为 1GB（默认）及 500MB 时，`EXPLAIN ANALYZE` 的返回结果（当内存配额设为 500MB 时，磁盘用作临时存储区）：
```

File: explain-joins.md [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
++-----------------------------+-----------+---------+-----------+---------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------+-----------------------+----------------------+
+5 rows in set (0.98 sec)
+```
+
+### 配置
+
+Hash Join 算法的性能受以下系统变量影响：
+
+* [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query) (默认值：1GB) - 如果某条查询的内存消耗超出了配额，TiDB 会尝试将 Hash Join 的 Build 端移到磁盘上以节省内存。
+* [`tidb_hash_join_concurrency`](/system-variables.md#tidb_hash_join_concurrency) (默认值：`5`) - 可以并发执行的 Hash Join 任务数量。
```
Comment (TomShawn) [inline explain-joins.md:270]:
```suggestion
* [`tidb_mem_quota_query`](/system-variables.md#tidb_mem_quota_query)（默认值：1GB) - 如果某条查询的内存消耗超出了配额，TiDB 会尝试将 Hash Join 的 Build 端移到磁盘上以节省内存。
* [`tidb_hash_join_concurrency`](/system-variables.md#tidb_hash_join_concurrency)（默认值：`5`）- 可以并发执行的 Hash Join 任务数量。
```

File: explain-joins.md [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+
+## Merge Join
+
+Merge Join 是一种特殊的 Join 算法。当两个关联表要 Join 的字段需要按排好的顺序读取时，适用 Merge Join 算法。由于 Build 端和 Probe 端的数据都会读取，这种算法的 Join 操作是流式的，类似“拉链式合并”的高效版。Merge Join 占用的内存要远低于 Hash Join，但 Merge Join 不能并发执行。
+
+下面给出了一个例子：
```
Comment (TomShawn) [inline explain-joins.md:276]:
```suggestion
下面是一个使用 Merge Join 的例子：
```

File: explain-joins.md [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+| MergeJoin_7                 | 142020.00 | root      |               | inner join, left key:test.t1.id, right key:test.t2.id |
+| ├─TableReader_12(Build)     | 180000.00 | root      |               | data:TableFullScan_11                                 |
+| │ └─TableFullScan_11        | 180000.00 | cop[tikv] | table:t2      | keep order:true                                       |
+| └─TableReader_10(Probe)     | 142020.00 | root      |               | data:TableFullScan_9                                  |
+|   └─TableFullScan_9         | 142020.00 | cop[tikv] | table:t1      | keep order:true                                       |
++-----------------------------+-----------+-----------+---------------+-------------------------------------------------------+
+5 rows in set (0.00 sec)
+```
+
+TiDB 会按照以下操作步骤执行 Merge Join 算子：
```
Comment (TomShawn) [inline explain-joins.md:297]:
```suggestion
TiDB 会按照以下顺序执行 Merge Join 算子：
```

File: explain-joins.md [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+
+1. 从 Build 端把一个 Join Group 的数据全部读取到内存中。
+2. 读取 Probe 端的数据。
+3. 将 Probe 端的每行数据与 Build 端的一个完整 Join Group 比较，依次查看是否匹配（除了满足等值条件以外，还有其他非等值条件，这里的 “匹配” 主要是指查看是否满足非等值条件）。Join Group 指的是所有 Join Key 上值相同的数据。
```
Comment (TomShawn) [inline explain-joins.md:301]:
```suggestion
3. 将 Probe 端的每行数据与 Build 端的一个完整 Join Group 比较，依次查看是否匹配（除了满足等值条件以外，还有其他非等值条件，这里的“匹配”主要是指查看是否满足非等值条件）。Join Group 指的是所有 Join Key 上值相同的数据。
```

**Review by TomShawn — COMMENTED** [review]
Very minor comments on the format.
Great translation!

**Comment by TomShawn** [comment]
LGTM

**Review by lzmhhh123 — APPROVED** [review]
LGTM

**Review by TomShawn — APPROVED** [review]

**Pull request merged**